crates/holochain_types/src/chain.rs

```diff
@@ -46,6 +48,64 @@ pub struct ChainFilterIter {
     end: bool,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// A [`ChainFilter`] with the action sequences for the
+/// starting position and any `until` hashes.
+pub struct ChainFilterRange {
+    /// The filter for for this chain.
+    filter: ChainFilter,
+    /// The start of this range is the end of
+    /// the filter iterator.
+    /// The end of this range is the sequence of
+    /// the starting position hash.
+    range: RangeInclusive<u32>,
+    /// The start of the ranges type.
+    range_start_type: RangeStartType,
```

Comment:
It's kind of confusing that the start and end are reversed here wrt to the original filter. Maybe the original filter can use "top" and "bottom" instead of "start" and "end" so that these two don't contradict each other?

Also is it redundant to include the `filter` in this struct? Seems like this struct is just a different representation of the `filter` which produces it.

Reply:
Yeh I like the idea of top instead of position. As for redundancy this structure is purely extra information on top of the filter. The filter does not know the it's range as it does not know the positions of the `until` hashes.